File: src/java/nz/co/searchwellington/tagging/TagHintAutoTagger.java
package nz.co.searchwellington.tagging;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

import nz.co.searchwellington.model.Resource;
import nz.co.searchwellington.model.Tag;
import nz.co.searchwellington.repositories.ResourceRepository;

import org.apache.log4j.Logger;

public class TagHintAutoTagger {

	Logger log = Logger.getLogger(TagHintAutoTagger.class);
	private ResourceRepository resourceDAO;

	public TagHintAutoTagger(ResourceRepository resourceDAO) {
		this.resourceDAO = resourceDAO;
	}

	public Set<Tag> suggestTags(Resource resource) {
		Set<Tag> suggestedTags = new HashSet<Tag>();

		List<Tag> tags = resourceDAO.getAllTags();
		for (Tag tag : tags) {
			String tagHints = tag.getAutotagHints();
			if (tagHints != null && !tagHints.trim().equals("")) {
				String[] hints = tagHints.split(",");
				if (hints.length > 0) {
					for (int i = 0; i < hints.length; i++) {
						String hint = hints[i].trim();
						if (checkForMatch(resource, hint)) {
							log.info("Suggesting tag '" + tag.getDisplayName() + "' for resource: " + resource.getName());
							suggestedTags.add(tag);
						}
					}
				}
			}
		}
		return suggestedTags;

	}

	private boolean checkForMatch(Resource resource, String hint) {
		if (!hint.trim().equals("")) {
			boolean headlineMatchesHint = resource.getName().toLowerCase().contains(hint.toLowerCase());
			boolean BodyMatchesTag = resource.getDescription().toLowerCase().contains(hint.toLowerCase());
			return headlineMatchesHint || BodyMatchesTag;
		}
		return false;
	}

}

File: src/main/scala/nz/co/searchwellington/tagging/TagHintAutoTagger.scala
package nz.co.searchwellington.tagging

import nz.co.searchwellington.feeds.whakaoko.model.Category
import nz.co.searchwellington.model.{Resource, Tag}
import nz.co.searchwellington.repositories.TagDAO
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component

import java.util.StringTokenizer
import scala.collection.mutable
import scala.collection.mutable.ListBuffer
import scala.concurrent.{ExecutionContext, Future}

@Component
class TagHintAutoTagger @Autowired()(tagDAO: TagDAO) {

  def suggestTags(resource: Resource)(implicit ec: ExecutionContext): Future[Set[Tag]] = {
    val resourceContent = resource.title + " " + resource.description
    val resourceTokens = tokenise(resourceContent.toLowerCase())

    tagDAO.getAllTags.map { allTags =>
      allTags.filter(tag => matches(tag.hints, resourceTokens)).toSet
    }
  }

  def suggestFeedCategoryTags(feedItemCategories: Seq[Category])(implicit ec: ExecutionContext): Future[Set[Tag]] = {
    if (feedItemCategories.nonEmpty) {
      // Given all tags look for autohints with vaguely match the categories presented
      val categoryTokens = feedItemCategories.map(_.value.toLowerCase()).map(tokenise)

      tagDAO.getAllTags.map { allTags =>
        allTags.filter { tag =>
          categoryTokens.exists { categoryTokens =>
            matches(tag.hints, categoryTokens)
          }
        }.toSet
      }
    } else {
      Future.successful(Set.empty)
    }
  }

  private def matches(hints: Seq[String], resourceTokens: Seq[String]): Boolean = {
    hints.exists { hint =>
      val hintTokens = tokenise(hint.toLowerCase)
      resourceTokens.containsSlice(hintTokens)
    }
  }

  private def tokenise(resourceContent: String): Seq[String] = {
    val standardTokenizer = new StringTokenizer(resourceContent)
    val tokens: mutable.ListBuffer[String] = ListBuffer.empty
    val iterator = standardTokenizer.asIterator()
    while (iterator.hasNext) {
      val value = iterator.next().asInstanceOf[String]
      tokens += value
    }
    tokens.toSeq
  }


}
